cdn/sign_url.rb

```diff
@@ -0,0 +1,64 @@
+# Copyright 2018 Google, Inc
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def signed_url url:, key_name:, key:, expiration:
+  # [START signed_url]
+  # url        = "URL of the endpoint served by Cloud CDN"
+  # key_name   = "Name of the signing key added to the Google Cloud Storage bucket or service"
+  # key        = "Signing key as urlsafe base64 encoded string"
+  # expiration = Time.now + 60*60 # Signed URL expires an hour from now.
+
+  require "base64"
+  require "openssl"
+  require "time"
+
+  # Decode the URL safe base64 encode key
+  decoded_key = Base64.urlsafe_decode64 key
+
+  # Get UTC time in seconds
+  expiration_utc = expiration.utc.to_i
+
+  # Determine which seperator makes sense given a URL
+  seperator = "?"
+  seperator = "&" if url.include? '?'
+
+  # Concatenate url with expected query parameters Expires and KeyName
+  url = "#{url}#{seperator}Expires=#{expiration_utc}&KeyName=#{key_name}"
+
+  # Sign the url using the key and url safe base64 encode the signature
+  signature         = OpenSSL::HMAC.digest "SHA1", decoded_key, url
+  encoded_signature = Base64.urlsafe_encode64 signature
+
+  puts "Signed URL: #{url}&Signature=#{encoded_signature}"
```

Comment:
I don't know ruby, but it looks like this method isn't returning a URL, it's rather printing it to the screen, is this correct? This method should just return a string.

Reply:
For other Ruby samples this is the way they introduce values for sample code.